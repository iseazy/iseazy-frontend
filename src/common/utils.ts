export function getFormattedTimeDifference(fromDate: number, toDate: number = Date.now()): string {
    const millis = toDate -fromDate;
    const minutes = Math.floor(millis / 60000);
    const seconds = ((millis % 60000) / 1000).toFixed(0);
    let gameDuration = '';
    
    if (minutes > 0) {
        gameDuration += `${minutes}min `;
    }

    return `${gameDuration}${seconds}s`;
}

export function shuffleList<T>(items: Array<T>): Array<T> {
    return items.sort(() => Math.random() - 0.5);
}
