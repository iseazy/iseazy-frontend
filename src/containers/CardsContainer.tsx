import { useContext, useState, useEffect, useRef } from 'react';
import Card from '../components/Card';
import styled from 'styled-components/macro'
import AppContext, { Card as ICard } from '../hooks/AppContext';

const CardsContainerStyled = styled.div`
    top: 50%;
    margin: 0 auto;
    width: 900px;
    display: grid;
    grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr;
    grid-template-rows: 1fr 1fr 1fr;
    gap: 15px 5px;
`;
const CardsContainer = () => {
    const { state, dispatch } = useContext(AppContext);
    const [cardsSelected, setCardsSelected] = useState<ICard[]>([])
    const { cards, flipped } = state;

    const handleClickOnCard = (value: ICard) => {
        if (cardsSelected.length < 2) {
            setCardsSelected([...cardsSelected, value])
        } else {
            setCardsSelected([value])
        }
    }

    const cardIsFlipped = (card: ICard) => flipped.some((c: ICard) => c?.index === card.index) || cardsSelected.some((c: ICard) => c.id === card.id);

    const wonConditionAllCardsFlipped = (flipped: ICard[], cards: ICard[]) => cards.every((card) => flipped.some((flippedCardPair) => flippedCardPair.index === card.index))

    useEffect(() => {
        if (wonConditionAllCardsFlipped(flipped, cards)) {
            dispatch({ type: 'won' })
        }
    }, [flipped])

    useEffect(() => {
        if (cardsSelected.length >= 2) {
            if (cardsSelected[0].index === cardsSelected[1].index) {
                dispatch({ type: 'flip', value: cardsSelected[1] })
                setCardsSelected([])
            }
            const timer = setTimeout(() => {
                setCardsSelected([]);
            }, 500);
            return () => clearTimeout(timer);
        }
    }, [cardsSelected])

    return (
        <CardsContainerStyled>
            {cards.map((i, index) => <Card
                onClick={() => handleClickOnCard(i)}
                shown={cardIsFlipped(i)}
                index={i.index}
                text={index}
                key={index} />)}
        </CardsContainerStyled>
    )
}

export default CardsContainer
