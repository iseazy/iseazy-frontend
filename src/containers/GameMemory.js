import { useState, useEffect } from "react";
import Card from "../components/Card";
import Timer from "../components/Timer";
import ModalEnd from "../components/ModalEnd";

import "../css/game.css";

// TODO: metodo rapido para la optencion de las imagenes
function card(src, match = false, id = null) {
  this.src = src;
  this.match = match;
  this.id = id;
}

const cardsCollection = [
  new card("/img/1.png"),
  new card("/img/2.png"),
  new card("/img/3.png"),
  new card("/img/4.png"),
  new card("/img/5.png"),
  new card("/img/6.png"),
  new card("/img/7.png"),
  new card("/img/8.png"),
  new card("/img/9.png"),
];

export default function GameMemory({ handleEndGame }) {
  const [cards, setCards] = useState([]);
  const [firstChoice, setFirstChoice] = useState(null);
  const [secondChoice, setSecondChoice] = useState(null);

  const [timeRecord, setTimeRecord] = useState(null);
  const [gameOver, setGameOver] = useState(false);

  // Crear una nueva baraja aleatoria al inicio del juego
  const shuffleCards = () => {
    const shuffle = [...cardsCollection, ...cardsCollection]
      .sort(() => Math.random() - 0.5)
      .map((card) => ({ ...card, id: Math.random() }));

    setCards(shuffle);
  };
  useEffect(shuffleCards, []);

  // handle para manejar las selecciones de cartas del jugador
  const handleChoice = (card) => {
    firstChoice ? setSecondChoice(card) : setFirstChoice(card);
  };

  // reseteo de las 2 opciones elegidas por el jugador
  const resetChoice = () => {
    setFirstChoice(null);
    setSecondChoice(null);
  };

  // efecto que compara las cartas seleccionadas por el jugador y calcula si hacen match
  const compareCards = () => {
    if (firstChoice && secondChoice) {
      if (firstChoice?.src === secondChoice?.src) {
        setCards((prevCards) => {
          return prevCards.map((card) => {
            if (card.src === firstChoice.src) {
              return { ...card, match: true };
            } else {
              return card;
            }
          });
        });
        resetChoice();
      } else {
        // pequeÃ±o delay para que el jugador le pueda dar tiempo a ver las 2 cartas antes de darles la vuelta
        setTimeout(() => {
          resetChoice();
        }, 1000);
      }
    }
  };
  useEffect(compareCards, [firstChoice, secondChoice]);

  // Efecto encargado de determinar si se han encontrado todos los pares de cartas y que cambia el estado que lanzar el evento de fin de juego
  const launchEndGame = () => {
    const isAllPairsMatch = cards.find((card) => card.match === false);
    if (cards.length > 0 && !isAllPairsMatch) {
      setGameOver(true);
    }
  };
  useEffect(launchEndGame, [cards]);

  // hendle que setea el tiempo de juego del jugador
  const handleTimeRecord = (time) => {
    setTimeRecord(time);
  };

  // handle que finaliza el juego
  const handleGameOver = () => {
    handleEndGame();
  };

  return (
    <>
      <div data-testid="testid-timer-gameMemory" className="memory-card-grid">
        {cards.map((card) => (
          <Card
            card={card}
            handleChoice={handleChoice}
            flipped={
              card === firstChoice || card === secondChoice || card.match
            }
            comparationInProgress={secondChoice ? true : false}
            key={card.id}
          />
        ))}
      </div>

      <Timer gameOver={gameOver} handleTimeRecord={handleTimeRecord} />

      {gameOver ? (
        <ModalEnd timeRecord={timeRecord} handleGameOver={handleGameOver} />
      ) : (
        ""
      )}
    </>
  );
}
