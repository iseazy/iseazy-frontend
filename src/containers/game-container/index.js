export { GameContainer } from './game-container';
