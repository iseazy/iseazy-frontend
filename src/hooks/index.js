export { useGame } from './use-game';
