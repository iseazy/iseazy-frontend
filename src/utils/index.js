export {
    assignCardId,
    createCard,
    flipCardById,
    isCardClicked,
    isCardFlipped,
    isCardNotClicked,
    hasCardSharedId,
    unclickCard,
    unflipClickedCard
} from './card';

export {
    areClickedCardsPaired,
    createDeck,
    isDeckSolved
} from './deck';

export { isGameEnd, isGameIdle } from './game';

export { shuffleArray } from './suffle-array';

export { getFormattedTime } from './time';
