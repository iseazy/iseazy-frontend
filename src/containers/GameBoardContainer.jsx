import React, { useState, useContext, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import GameContext from '../contexts/GameContext'
import useTheme from '../hooks/useTheme'
import Card from '../components/card/Card'
import styled from 'styled-components';
import Timer from '../components/timer/Timer'
import FinishGameModal from '../components/modal/FinishGameModal';
import { ActionTypesScoreBoard } from '../constants/enums/actionTypesScoreBoard';
import { formatDateTime } from '../utils/utils';
import { ROUTE_HOME } from '../routes/routes';

const GameBoardGrid = styled.div`
  display: grid;
  grid-template-rows: repeat(3, 1fr);
  grid-template-columns: repeat(6, 1fr);
`;

const TimerContainer = styled.div`
  position: absolute;
  top: 0;
  left: 50%;
  transform: translateX(-50%);
  padding: 3rem;
`;

const GameBoardContainer = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate()
  const { onClickCard, time, onFinishGame, onResetGame, gameState: { boardState, shuffledImages, lockedCards, } } = useContext(GameContext)
  const { currentTheme } = useTheme()
  const [isModalOpen, setModalOpen] = useState(false);

  useEffect(() => {
    if (!shuffledImages.length) {
      navigate(`${ROUTE_HOME}`)
    }
  }, [])


  const onCloseModal = () => {
    setModalOpen(false)
    onResetGame()
  }

  const onClickRestartGame = () => {
    onCloseModal()
  }

  const handleFinishGame = () => {
    onFinishGame();
    const newScore = { date: formatDateTime(new Date()), score: time }
    dispatch({ type: ActionTypesScoreBoard.ADD_SCORE, payload: newScore });
    setModalOpen(true);
  }

  const handleCardClick = (cardIndex) => {
    const isGameFinished = onClickCard(cardIndex)
    if (isGameFinished) {
      setTimeout(() => {
        handleFinishGame()
      }, 800)
    }
  }

  return (
    <div>
      <TimerContainer>
        <Timer time={time} currentTheme={currentTheme} />
      </TimerContainer>
      <GameBoardGrid>
        {shuffledImages.map((image, index) => (
          <Card
            key={index}
            flipped={boardState[index]}
            locked={lockedCards.has(index)}
            imageSrc={image}
            number={index + 1}
            onClick={() => handleCardClick(index)}
          />
        ))}
      </GameBoardGrid>
      <FinishGameModal time={time} isModalOpen={isModalOpen} onClose={onCloseModal} onClickButton={onClickRestartGame} />
    </div>
  );
};

export default GameBoardContainer;
