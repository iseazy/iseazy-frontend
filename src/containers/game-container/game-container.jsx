import { GameEnd } from '../../components/game-end';
import { GameIntro } from '../../components/game-intro';
import { GamePlay } from '../../components/game-play';
import { useGame } from '../../hooks';
import { isGameEnd, isGameIdle } from '../../utils';

export const GameContainer = () => {
    const game = useGame();

    return (
        <div className="game-container">
            { isGameIdle(game)
                ? <GameIntro onClick={ game.start } />
                : <GamePlay cards={ game.cards } onClick={ game.clickCard } /> }
            { isGameEnd(game)
                ? <GameEnd time={ game.totalTime } onClick={ game.start } />
                : null }
        </div>
    );
};
