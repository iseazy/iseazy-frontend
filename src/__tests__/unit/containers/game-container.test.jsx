import { render } from '@testing-library/react';
import { GameContainer } from '../../../containers/game-container';
import { isGameEnd, isGameIdle } from '../../../utils';
import { useGame } from '../../../hooks';

jest.mock('../../../components/game-intro', () => ({
    GameIntro: () => 'GameIntro'
}));
jest.mock('../../../components/game-play', () => ({
    GamePlay: () => 'GamePlay'
}));
jest.mock('../../../components/game-end', () => ({
    GameEnd: () => 'GameEnd'
}));
jest.mock('../../../utils', () => ({
    isGameEnd: jest.fn(),
    isGameIdle: jest.fn()
}));
jest.mock('../../../hooks', () => ({
    useGame: jest.fn()
}));

describe('<GameContainer />', () => {
    beforeEach(() => {
        isGameEnd.mockReturnValue(false);
        isGameIdle.mockReturnValue(false);
        useGame.mockReturnValue({
            cards: [],
            totalTime: 123,
            start: jest.fn(),
            clickCard: jest.fn()
        });
    });

    it('should render game play screen', () => {
        const { asFragment } = render (<GameContainer />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render game intro screen', () => {
        isGameIdle.mockReturnValueOnce(true);
        const { asFragment } = render(<GameContainer />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render game end screen in top of game play screen', () => {
        isGameEnd.mockReturnValueOnce(true);
        const { asFragment } = render(<GameContainer />);

        expect(asFragment()).toMatchSnapshot();
    });
});
