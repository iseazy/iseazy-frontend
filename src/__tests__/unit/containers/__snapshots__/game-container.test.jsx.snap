// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameContainer /> should render game end screen in top of game play screen 1`] = `
<DocumentFragment>
  <div
    class="game-container"
  >
    GamePlayGameEnd
  </div>
</DocumentFragment>
`;

exports[`<GameContainer /> should render game intro screen 1`] = `
<DocumentFragment>
  <div
    class="game-container"
  >
    GameIntro
  </div>
</DocumentFragment>
`;

exports[`<GameContainer /> should render game play screen 1`] = `
<DocumentFragment>
  <div
    class="game-container"
  >
    GamePlay
  </div>
</DocumentFragment>
`;
